tests/common/snappi_tests/snappi_fixtures.py

```diff
@@ -1307,3 +1307,208 @@ def check_fabric_counters(duthost):
                               format(crc_errors, duthost.hostname, val_list[0], val_list[1]))
                 pytest_assert(fec_uncor_err == 0, 'Forward Uncorrectable errors:{} for DUT:{}, ASIC:{}, Port:{}'.
                               format(fec_uncor_err, duthost.hostname, val_list[0], val_list[1]))
+
+
+def snappi_port_selection(tmp_snappi_port_list, rx_port_count, tx_port_count, mixed_speed=None):
+    '''
+    Dynamic selection of the DUT ports for the test.
+    Selects ports for three test combinations:
+            - Single line-card single asic
+            - Single line-card multiple asic
+            - Multiple line-card.
+    Args:
+        port_list(list): list of the ports for each of selected dut.
+        rx_port_count(int): count of Rx DUT ports required for the test.
+        tx_port_count(int): count of Tx DUT ports requred for the test.
+    Returns:
+        snappi_ports(dict): Dictionary with interface-speed and line-card-combo being primary keys.
+        Example: {'100':{'single-linecard-single-asic':{ports}, 'single-linecard-multiple-asic':{ports}}}
+
+    '''
+
+    if (not mixed_speed):
+        # Creating list of all interface speeds from selected ports.
+        port_speed_list = []
+        for item in tmp_snappi_port_list:
+            if (int(item['speed'])/1000) not in port_speed_list:
+                port_speed_list.append(int(item['speed'])/1000)
+
+        port_list = {}
+        # Repeating loop for speed_types
+        for port_speed in port_speed_list:
+            new_list = []
+            # Selecting ports matching the port_speed
+            for item in tmp_snappi_port_list:
+                if (int(item['speed']) == (port_speed * 1000)):
+                    new_list.append(item)
+
+            # Creating dictionary f{hostname}{asic_val}
+            # f[hostname]['asic'] should contain associated elements.
+            f = {}
+            for item in new_list:
+                hostname = item['peer_device']
+                asic = item['asic_value']
+                if hostname not in f:
+                    f[hostname] = {}
+                if asic not in f[hostname]:
+                    f[hostname][asic] = []
+                f[hostname][asic].append(item)
+
+            total_ports = tx_port_count + rx_port_count
+
+            # Initializing dictionary port_list{speed}{line-card-asic-combo}
+            # example port_list['100']['single_linecard_single_asic']
+
+            # for 'single-linecard-single-asic'
```

Comment:
Hi @amitpawar12 

This code works for me. Just a few improvement suggestions which could help improving the readability.

I notice that you're separating into 3 cases:
1. for 'single-linecard-single-asic'
2. for 'single_linecard_multiple_asic'
3. for 'multiple linecard, multiple ASIC'


Could we have some if statement to separate these logic instead of letting the code goes through all? It will be easier to debug if we need to modify later.

A suggestion could be:

```python
is_single_linecard_single_asic = len(f) == 1 and len(f[0].keys()) == 1
if (is_single_linecard_single_asic):
   # logic for single_linecard_single_asic
   
is_single_linecard_multi_asic = len(f) == 1 and len(f[0].keys()) >= 1
 
if (is_single_linecard_multi_asic):
   # logic for is_single_linecard_multi_asic
   
```

cc @sdszhang